### PR TITLE
Map

### LINKED ISSUE #27: Implementing Map/Dict functionality with ARgorithmToolkit objects.
To implement a lot of algorithms, a dict or hashmap maybe a necessary Data-structure. It'll require some form of generics or interface classes to use as key-value pairs, to ensure that the correct data is being written into the states, and object references aren't lost.

### PR DESCRIPTION
**Pull request by yatharthmathur**
Added Map and Set data-structures
Improved overall design to accommodate them.
resolves #27

### CODE REVIEW

**Comment by alanpjohn** [comment]
#29 has to be closed before merging `map` with `master`

File: .vscode/settings.json
Comment (alanpjohn) [inline .vscode/settings.json:6]: can we keep this to false
i prefer using pylint from the command line it allows me to filter types of linting errors and more control
when VScode does it, it just highlights all my code and lowers my self-esteem which in turn lowers my productivity

**Review by alanpjohn — APPROVED** [review]

File: .vscode/settings.json
Comment (yatharthmathur) [reply .vscode/settings.json:6]: cool, it switched on automatically through my VSCode settings

**Comment by alanpjohn** [comment]
@yatharthmathur  just test the merge once
need to add some docs for server part as well with [changes](https://github.com/ARgorithm/Server/pull/14) set to be pushed
will merge after that

**Comment by alanpjohn** [comment]
> @yatharthmathur just test the merge once
> need to add some docs for server part as well with [changes](https://github.com/ARgorithm/Server/pull/14) set to be pushed
> will merge after that

Done my changes, just check everything from your side and we can merge it

**Pull request merged**
